mirror_builder/sources.py

```diff
@@ -0,0 +1,43 @@
+import logging
+
+import resolvelib
+from packaging.requirements import Requirement
+
+from . import overrides, resolve_and_download
+
+logger = logging.getLogger(__name__)
+
+
+def download_source(ctx, req):
+    logger.info('downloading source for %s', req)
+    r = Requirement(req)
+    downloader = overrides.find_override_method(r.name, 'download_source')
+    if not downloader:
+        downloader = _default_download_source
+    source_filename = downloader(ctx, req)
+    logger.info('downloaded source for %s to %s', req, source_filename)
+    return source_filename
```

Comment:
I had envisioned the main entry point to downloading source returning the path to where the expanded archive or git clone was written. That's easier to make common between various approaches I thought we might need to support (sdist, GitHub release tarball, git clone, etc.), and all of the rest of the higher level code should be able to work with a directory. I hadn't gotten as far as exploring what that API would do to the rest of the code, and I know we do rely on the sdist archive name quite a bit.

This function would then return that directory name, and each implementation would be responsible for unpacking the archive, if necessary.

We _might_ need to define a class to abstract different types of source structures, but I was hoping to not go that far until it was obviously needed.

Reply:
Ok, there's a prepare_source() hook point now